extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/db/SQLBasicSecurityStorageConnector.java

```diff
@@ -0,0 +1,1002 @@
+                        "SELECT user_name FROM %1$s WHERE user_name = :userName",
+                        USER_CREDENTIALS
+                    )
+                )
+                .bind("userName", userName)
+                .first();
+
+            int iterations = BasicAuthUtils.KEY_ITERATIONS;
+            byte[] salt = BasicAuthUtils.generateSalt();
+            byte[] hash = BasicAuthUtils.hashPassword(password, salt, iterations);
+
+            if (existingMapping == null) {
+              handle.createStatement(
+                  StringUtils.format(
+                      "INSERT INTO %1$s (user_name, salt, hash, iterations) " +
+                      "VALUES (:userName, :salt, :hash, :iterations)",
+                      USER_CREDENTIALS
+                  )
+              )
+                    .bind("userName", userName)
+                    .bind("salt", salt)
+                    .bind("hash", hash)
+                    .bind("iterations", iterations)
+                    .execute();
+            } else {
+              handle.createStatement(
+                  StringUtils.format(
+                      "UPDATE %1$s SET " +
+                      "salt = :salt, " +
+                      "hash = :hash, " +
+                      "iterations = :iterations " +
+                      "WHERE user_name = :userName",
+                      USER_CREDENTIALS
+                  )
+              )
+                    .bind("userName", userName)
+                    .bind("salt", salt)
+                    .bind("hash", hash)
+                    .bind("iterations", iterations)
+                    .execute();
+            }
+
+            return null;
+          }
+        }
+    );
+  }
+
+  @Override
+  public boolean checkCredentials(String userName, char[] password)
+  {
+    return getDBI().inTransaction(
+        new TransactionCallback<Boolean>()
+        {
+          @Override
+          public Boolean inTransaction(Handle handle, TransactionStatus transactionStatus) throws Exception
+          {
+            Map<String, Object> credentials = handle
+                .createQuery(
+                    StringUtils.format(
+                        "SELECT * FROM %1$s WHERE user_name = :userName",
+                        USER_CREDENTIALS
+                    )
+                )
+                .bind("userName", userName)
+                .map(credsMapper)
+                .first();
+
+            if (credentials == null) {
+              return false;
+            }
+
+            byte[] dbSalt = (byte[]) credentials.get("salt");
+            byte[] dbHash = (byte[]) credentials.get("hash");
+            int iterations = (int) credentials.get("iterations");
+
+            byte[] hash = BasicAuthUtils.hashPassword(password, dbSalt, iterations);
+
+            return Arrays.equals(dbHash, hash);
+          }
+        }
+    );
+  }
+
+  private int getUserCountInTransaction(Handle handle, String userName)
```

Comment:
What does the "InTransaction" suffix mean in this and below methods?

Reply:
It's referring to how they're called from within the DB transactions, for example:

```
    getDBI().inTransaction(
        new TransactionCallback<Void>()
        {
          @Override
          public Void inTransaction(Handle handle, TransactionStatus transactionStatus) throws Exception
          {
```

Reply:
Thanks for the explanation. I was confused with its name. It sounds to me like getUserCount() is executed in a transaction. I think just getUserCount() is fine.

Reply:
Cool, I changed the names to remove "inTransaction"